File: CSES/Building/soln.cpp
/*
 * written by Yaqi Zhang (zhang623@wisc.edu)
 * University of Wisconsin-Madison
 * 2019
 */

#include<bits/stdc++.h>

using namespace std;

const int INF = 0x3f3f3f3f;

int maxArea(vector<int> & hs) {
    stack<int> stk;
    int mx = 0, n = hs.size();
    for(int i = 0; i < n; ++i) {
        while(!stk.empty() && hs[i] < hs[stk.top()]) {
            int h = hs[stk.top()];
            stk.pop();
            int pre = stk.empty() ? -1 : stk.top();
            int w = i - pre - 1;
            mx = max(mx, w * h);
        }
        stk.push(i);
    }
    return mx;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(0);
    int m, n;
    cin >> m >> n;
    vector<string> grid(m);
    for(int i = 0; i < m; ++i) cin >> grid[i];
    vector<int> dp(n + 1, 0);
    int mx = 0;
    for(int i = 0; i < m; ++i) {
        for(int j = 0; j < n; ++j) {
            dp[j] = grid[i][j] == '*' ? 0 : dp[j] + 1;
        }
        mx = max(mx, maxArea(dp));
    }
    cout << mx << endl;
    return 0;
}

File: CSES/Building/soln.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
##################################
# University of Wisconsin-Madison
# Author: Yaqi Zhang
##################################
# This module contains ...
##################################

# standard library
import sys

def max_rectangle(hs):
    "compute the maximum rectangle of a histogram"
    hs.append(-1)
    stack = [-1]
    mx = 0
    for i, h in enumerate(hs):
        while h < hs[stack[-1]]:
            height = hs[stack.pop()]
            width = i - stack[-1] - 1
            mx = max(mx, width * height)
        stack.append(i)
    hs.pop()
    return mx

def main():
    r, c = map(int, input().split())
    grid = []
    for i in range(r):
        grid.append(input())

    hs = [1 if val == '.' else 0 for val in grid[0]]
    mx = max_rectangle(hs)
    for i in range(1, r):
        for j in range(c):
            if grid[i][j] == '*':
                hs[j] = 0
            else:
                hs[j] += 1
        mx = max(mx, max_rectangle(hs))
    print(mx)

if __name__ == "__main__":
    main()
